src/services/bunker.py

```diff
@@ -1,15 +1,413 @@
+
+        return vault_withdrawals
+
+    def _get_lido_validators_with_others(
+        self, blockstamp: BlockStamp
+    ) -> tuple[list[Validator], list[LidoKey], list[LidoValidator]]:
+        lido_keys = self.w3.kac.get_all_lido_keys(blockstamp)
+        validators = self.w3.cc.get_validators(blockstamp.state_root)
+        lido_validators = self.w3.lido_validators.filter_lido_validators(lido_keys, validators)
+
+        return validators, lido_keys, lido_validators
+
+    @staticmethod
+    def _calculate_total_effective_balance(all_validators: list[Validator]) -> Gwei:
+        """
+        Calculates total balance of all active validators in network
+        """
+        total_effective_balance = 0
+
+        for v in all_validators:
+            if v.status in [
+                ValidatorStatus.ACTIVE_ONGOING,
+                ValidatorStatus.ACTIVE_EXITING,
+                ValidatorStatus.ACTIVE_SLASHED
+            ]:
+                total_effective_balance += int(v.validator.effective_balance)
+
+        return Gwei(total_effective_balance)
+
+    @staticmethod
+    def _not_withdrawn_slashed_validators(all_validators: list[Validator], ref_epoch: EpochNumber) -> list[Validator]:
+        """
+        Get all slashed validators, who are not withdrawn yet
+        """
+        slashed_validators = []
+
+        for validator in all_validators:
+            v = validator.validator
+            if v.slashed and int(v.withdrawable_epoch) > ref_epoch:
+                slashed_validators.append(validator)
+
+        return slashed_validators
+
+    @staticmethod
+    def _get_per_epoch_buckets(
+        all_slashed_validators: list[Validator], ref_epoch: EpochNumber
+    ) -> dict[EpochNumber, list[Validator]]:
+        """
+        Fill per_epoch_buckets by possible slashed epochs
+        It detects slashing epoch range for validator
+        If difference between validator's withdrawable epoch and exit epoch is greater enough,
+        then we can be sure that validator was slashed in particular epoch
+        Otherwise, we can only assume that validator was slashed in epochs range
+        due because its exit epoch shifted because of huge exit queue
+        Read more here: https://hackmd.io/@lido/r1Qkkiv3j
+        """
+        per_epoch_buckets = defaultdict(list[Validator])
+        for validator in all_slashed_validators:
+            v = validator.validator
+            if not v.slashed:
+                raise Exception("Validator should be slashed to detect slashing epoch range")
+            possible_slashed_epoch = int(v.withdrawable_epoch) - EPOCHS_PER_SLASHINGS_VECTOR
+            if int(v.withdrawable_epoch) - int(v.exit_epoch) > MIN_VALIDATOR_WITHDRAWABILITY_DELAY:
+                per_epoch_buckets[possible_slashed_epoch].append(validator)
+                continue
+            else:
+                for epoch in range(possible_slashed_epoch, ref_epoch + 1):
```

Comment:
```
               <-------36d------>
      [-------?------------------] slashing
      se     pse                 we
====================|=====================> time
                ref_epoch
```
It shows a slashing, the beginning of which cannot be determined by validator state

`se` - slashed epoch
`pse` - possible slashed epoch (we - 36d)
`we` - withdrawable epoch

We are interested in the following range (`x` for mandatory and `o` for optional):
```
               <-------36d------>
      [-------?------------------] slashing
      se     pse                 we
ooooooxxxxxxxxx=====|=====================> time
                ref_epoch
```

Reply:
I guess the range should be [ref_epoch - 36d, pse]

Reply:
Yes, sounds like the right approach. Done